### PR TITLE
feat: add downcast_owned_builder

### PR DESCRIPTION
**Pull request by ariesdevil**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

Add `try_upcast_column_builder` and `try_downcast_owned_builder` for ValueType, so the types like `NullableType` and `ArrayType` can downcast builder.

- Closes #issue

- - -
This change is [Reviewable](https://reviewable.io/reviews/datafuselabs/databend/13765)

### CODE REVIEW

**Review by sundy-li — APPROVED** [review]

File: src/query/expression/src/types/decimal.rs [hunk 1]
```diff
@@ -89,6 +89,14 @@ impl<Num: Decimal> ValueType for DecimalType<Num> {
         Num::try_downcast_builder(builder)
     }
 
+    fn try_downcast_owned_builder(builder: ColumnBuilder) -> Option<Self::ColumnBuilder> {
+        Num::try_downcast_owned_builder(builder)
+    }
+
+    fn try_upcast_column_builder(_builder: Self::ColumnBuilder) -> Option<ColumnBuilder> {
+        None
```
Comment (andylokandy) [inline src/query/expression/src/types/decimal.rs:97]: Why no to implement it?

File: src/query/functions/src/aggregates/aggregator_common.rs [hunk 2]
```diff
@@ -167,3 +168,24 @@ pub fn deserialize_state<T: serde::de::DeserializeOwned>(slice: &mut &[u8]) -> R
...
+        let closure_result = panic::catch_unwind(panic::AssertUnwindSafe(|| closure(old_t)));
+
+        match closure_result {
+            Ok(Ok(new_t)) => {
+                ptr::write(mut_ref, new_t);
+                Ok(())
+            }
+            Ok(Err(e)) => Err(e),
+            Err(_) => ::std::process::abort(),
+        }
```
Comment (andylokandy) [inline src/query/functions/src/aggregates/aggregator_common.rs:189]: Why not to use `take_mut`?

Comment (ariesdevil) [reply src/query/functions/src/aggregates/aggregator_common.rs:189]: We need return `Result<T>` in closure, so I copy the code and modify it.

Comment (andylokandy) [reply src/query/functions/src/aggregates/aggregator_common.rs:189]: fair enough. so common-base may be a better place for it.

File: src/query/expression/src/types/decimal.rs (hunk 1 above)
Comment (ariesdevil) [reply src/query/expression/src/types/decimal.rs:97]: I'm trying to refactor it, but there are huge changes that need to be made, so for now, leave it for future refactor

File: src/query/functions/src/aggregates/aggregator_common.rs (hunk 2 above)
Comment (ariesdevil) [reply src/query/functions/src/aggregates/aggregator_common.rs:189]: Let's move it to common-base

File: src/query/expression/src/types/number.rs [hunk 3]
```diff
@@ -908,6 +956,17 @@ impl Number for u64 {
         builder.as_u_int64_mut()
     }
 
+    fn try_downcast_owned_builder(builder: NumberColumnBuilder) -> Option<Vec<Self>> {
+        match builder {
+            NumberColumnBuilder::UInt64(b) => Some(b),
+            _ => unreachable!(),
```
Comment (andylokandy) [inline src/query/expression/src/types/number.rs:962]: Why not return None here?

Comment (ariesdevil) [reply src/query/expression/src/types/number.rs:962]: Yup, should be None

**Pull request merged**